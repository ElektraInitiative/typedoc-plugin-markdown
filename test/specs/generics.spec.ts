import * as Handlebars from 'handlebars';
import { ProjectReflection, SignatureReflection } from 'typedoc';

describe(`Generics:`, () => {
  let project: ProjectReflection;
  let partial: Handlebars.TemplateDelegate;
  let declarationPartial: Handlebars.TemplateDelegate;
  let reflectionTemplate: Handlebars.TemplateDelegate;

  beforeAll(() => {
    partial = global.getPartial('member.signature');
    declarationPartial = global.getPartial('member.declaration');
  });

  beforeEach(async () => {
    project = await global.bootstrap(['generics.ts']);
    global.stubPartials([
      'comment',
      'member.signature',
      'members',
      'member.sources',
    ]);
    global.stubHelpers(['toc', 'breadcrumbs', 'hierarchy', 'returns']);
    reflectionTemplate = global.getTemplate('reflection');
  });

  test(`should compile class with type params`, () => {
    expect(
      global.compileTemplate(reflectionTemplate, {
        model: project.getChildByName('ClassWithTypeParams'),
        project: project,
      }),
    ).toMatchSnapshot();
  });

  test(`should compile function with a simple type param'`, () => {
    expect(
      global.compileTemplate(
        partial,
        (project.getChildByName('functionWithTypeParam') as any)
          .signatures[0] as SignatureReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile function with complex type params'`, () => {
    expect(
      global.compileTemplate(
        partial,
        (project.getChildByName('functionWithTypeParams') as any)
          .signatures[0] as SignatureReflection,
      ),
    ).toMatchSnapshot();
  });

  test(`should compile type with nested generics'`, () => {
    expect(
      global.compileTemplate(
        declarationPartial,
        project.getChildByName('nestedGenerics'),
      ),
    ).toMatchSnapshot();
  });

  test(`should compile generics with defaults'`, () => {
    expect(
      global.compileTemplate(
        declarationPartial,
        (project.getChildByName('genericsWithDefaults') as any)
          .signatures[0] as SignatureReflection,
      ),
    ).toMatchSnapshot();
  });
});
