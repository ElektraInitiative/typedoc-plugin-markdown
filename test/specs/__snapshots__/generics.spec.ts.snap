// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generics: should compile class with type params 1`] = `
"[helper: breadcrumbs]

# Class: ClassWithTypeParams<T, V\\>

[partial: comment]

## Type parameters

| Name |
| :------ |
| \`T\` |
| \`V\` |

[helper: toc]

[partial: members]
"
`;

exports[`Generics: should compile function with a simple type param' 1`] = `
"▸ **functionWithTypeParam**<\`A\`\\>(): \`boolean\`

[partial: comment]

##### Type parameters

| Name |
| :------ |
| \`A\` |

##### Returns

\`boolean\`
"
`;

exports[`Generics: should compile function with complex type params' 1`] = `
"▸ **functionWithTypeParams**<\`A\`, \`B\`, \`C\`\\>(): \`boolean\`

[partial: comment]

##### Type parameters

| Name | Type | Description |
| :------ | :------ | :------ |
| \`A\` | extends [\`ClassWithTypeParams\`](ClassWithTypeParams.md)<\`string\`, \`number\`, \`A\`\\> | Comment for type \`A\` |
| \`B\` | \`string\` \\| \`boolean\` | Comment for type \`B\` |
| \`C\` | \`string\` | - |

##### Returns

\`boolean\`

[helper: returns]
"
`;

exports[`Generics: should compile generics with defaults' 1`] = `
"▸ **genericsWithDefaults**: \`void\`

[partial: comment]

#### Type parameters

| Name | Type |
| :------ | :------ |
| \`Type\` | extends \`boolean\` = \`boolean\` |

[partial: member.sources]
"
`;

exports[`Generics: should compile type with nested generics' 1`] = `
"Ƭ **nestedGenerics**: [\`Generic1\`](../modules.md#generic1)<[\`Generic2\`](../modules.md#generic2)<[\`Generic3\`](../modules.md#generic3)<\`string\`\\>\\>\\>

[partial: comment]

[partial: member.sources]
"
`;
